src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs

```diff
@@ -1052,46 +1093,59 @@ public async Task TcpReceiveSendGetsCanceledByDispose(bool receiveOrSend)
+                    try
                     {
-                        Assert.Equal(SocketError.OperationAborted, localSocketError);
-                    }
+                        if (UsesApm)
+                        {
+                            Assert.Null(localSocketError);
+                            Assert.True(disposedException);
+                        }
+                        else if (UsesSync)
+                        {
+                            Assert.Equal(SocketError.ConnectionAborted, localSocketError);
+                        }
+                        else
+                        {
+                            Assert.Equal(SocketError.OperationAborted, localSocketError);
+                        }
 
-                    // On OSX, we're unable to unblock the on-going socket operations and
-                    // perform an abortive close.
-                    if (!(UsesSync && PlatformDetection.IsOSXLike))
-                    {
-                        SocketError? peerSocketError = null;
-                        var receiveBuffer = new ArraySegment<byte>(new byte[4096]);
-                        while (true)
+                        // On OSX, we're unable to unblock the on-going socket operations and
+                        // perform an abortive close.
+                        if (!(UsesSync && PlatformDetection.IsOSXLike))
                         {
-                            try
+                            SocketError? peerSocketError = null;
+                            var receiveBuffer = new ArraySegment<byte>(new byte[4096]);
+                            while (true)
                             {
-                                int received = await ReceiveAsync(socket2, receiveBuffer);
-                                if (received == 0)
+                                try
                                 {
+                                    int received = await ReceiveAsync(socket2, receiveBuffer);
+                                    if (received == 0)
+                                    {
+                                        break;
+                                    }
+                                }
+                                catch (SocketException se)
+                                {
+                                    peerSocketError = se.SocketErrorCode;
                                     break;
                                 }
                             }
-                            catch (SocketException se)
-                            {
-                                peerSocketError = se.SocketErrorCode;
-                                break;
-                            }
+                            Assert.Equal(SocketError.ConnectionReset, peerSocketError);
                         }
-                        Assert.Equal(SocketError.ConnectionReset, peerSocketError);
+                        break;
+                    }
+                    catch
```

Comment:
This catch in particular worries me because it's wrapping a bunch of code, including the ReceiveAsync call above, and by eating all exceptions we could be masking unexpected exceptions.

Seems like maybe we shouldn't use try/catch at all to trigger retries, but instead just look at the results and determine whether a retry is needed. That seems safest.

Reply:
@geoffkizer that happened here mostly was that `RetryHelper` got unrolled to a manual loop, we are not any more permissive than before the change. However, if we are this concerned about fatal exceptions triggering unwanted retries, we should be probably much more careful about using `RetryHelper` or stop using it in general.

My original suggestion was to actually extend it with a `FatalException` to escape retries (see https://github.com/dotnet/runtime/commit/1339f52abb60c4af73306cd3776cb969ed898762), but that idea got some pushback from @tmds.

Maybe we need to do it the other way around: define a `RetryException` or some sort of exception filter to detect the permissive cases and and fail on everything else?

@stephentoub any thoughts about this?

Reply:
My initial inclination is to add a filter func to the retry helper methods that decides whether to handle the exception or not, e.g.
https://github.com/dotnet/runtime/blob/c788fe1786ffc9295f1c469e506e26ff2ff4f94c/src/libraries/Common/tests/TestUtilities/System/RetryHelper.cs#L56
becomes
```C#
public static async Task ExecuteAsync(Func<Task> test, int maxAttempts = 5, Func<int, int> backoffFunc = null, Func<Exception, bool> filterFunc = null)
```
and
https://github.com/dotnet/runtime/blob/c788fe1786ffc9295f1c469e506e26ff2ff4f94c/src/libraries/Common/tests/TestUtilities/System/RetryHelper.cs#L77
becomes
```C#
catch (Exception e) when (filterFunc is null || filterFunc(e))
```

Reply:
@stephentoub I think without further sugar this API would only complicate call sites (test-specifc custom exception types and/or complicated filter methods), making readability even worse. If I could design `RetryHelper` from scratch, I would do something like this:
https://gist.github.com/antonfirsov/d5cf9e01b065e58420080d0e9c31e548

In this PR, I will go with the simplest possible fix now, and not touch `RetryHelper`, since it looks like we don't have consensus and sufficient attention span here, but I believe this is a question that's worth coming back to later.

Reply:
>  I think without further sugar this API would only complicate call sites

Every existing usage would be unchanged unless it actually wanted to restrict the retry conditions.  If you wanted to only retry RetryExceptions, you'd just add `IsRetryException` or `e =>  e is RetryException` to the call site.

So I'm not seeing the complication...?

Reply:
By "complication" I meant boilerplate code. If we introduce a `RetryException` inside `RetryHelper.cs` instead of doing it repeatedly in individual test libraries, and maybe also add the `TryAssert` utility I suggested, then the extra `e => e is RetryException` parameter would be the only boilerplate which is fine for me.

Reply:
I'm not clear on why any of that is necessary.  Are you saying in this example that you want to retry for some of the asserts but not for others?

Reply:
Yes. in this PR we want to retry on some assertions/exceptions, but not on others. We masked some fatal exceptions in socket tests by retrying. My concern is that we probably want to fail immediately on a bigger portion of asserts (or exceptions) than we thought before, but I'm not familiar with `RetryHelper` usages outside of socket tests.

Reply:
I'm specifically asking about asserts, though.  In this test,  for example,  for what do we want to retry and what do we not want to retry?

Reply:
In this particular test we want to retry on these asserts:
https://github.com/dotnet/runtime/blob/1da5006d99c0de404cdc2678b20bcf9cf9e62175/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs#L994-L1007

But not on this one (and not any other non-xunit exceptions):
https://github.com/dotnet/runtime/blob/1da5006d99c0de404cdc2678b20bcf9cf9e62175/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs#L972

Of course, with a simple filter func we could retry on `XunitException` and throw some arbitrary exception instead of the second assert. But this doesn't feel like an ideal solution, would be harder to follow such a test code if I was new to the repo.

Reply:
> But not on this one

Those four lines should just be:
```C#
await Task.WhenAny(disposeTask, receiveTask).WithTimeout(30000);
```

Reply:
Ok, I will give such a refactor a try, and see how the code will look and how will people react.

Reply:
Sounds good.  If it works well,  great.  If not,  we can explore other options.